[RFXCOM] New PacketType: IT

I have some Flamingo FA500 switches lying around that are supported by the RFXCOM device through the IT protocol. It would be great if the binding will support that protocol, too.

Product page:
http://www.elro.eu/en/products/cat/flamingo/electricite/wireless-control-flamingo/2-switches-with-remote-control6

RFXtrx User Guide (see: 2.2.2. Alphabetic list):
http://www.rfxcom.com/WebRoot/StoreNL2/Shops/78165469/MediaGallery/Downloads/RFXtrx_User_Guide.pdf

Reverse engineering for Arduino:
**rf433 code for ELRO Flamingo home devices (FA500)**
https://forum.arduino.cc/index.php?topic=201771.0
##

The FA500S does not have a blinking red light; I need to send an ON command within 10 seconds after plugging it in. I did that. Please find a more elaborate debug log.

_My personal guess would be a) that the protocol is not yet enabled on the RFXCOM device or b) I need a firmware upgrade?_

The thing is, I do not have access to a Windows machine to use their configuration app.

```
2016-12-24 06:43:26.393 [DEBUG] [.rfxcom.internal.RFXComBinding] - Received command: Virtual_FA500S_1 ON
2016-12-24 06:43:26.395 [DEBUG] [.rfxcom.internal.RFXComBinding] - Received command (item='Virtual_FA500S_1', state='ON', class='class org.openhab.core.library.types.OnOffType')
2016-12-24 06:43:26.397 [DEBUG] [.rfxcom.internal.RFXComBinding] - Transmitting data: 0A140F19000B5501010000
2016-12-24 06:43:26.446 [DEBUG] [.b.r.internal.RFXComConnection] - Data received:
Raw data = 0D01FF190053E2000C2701020000
 - Packet type = INTERFACE_MESSAGE
 - Seq number = 25
 - Sub type = UNKNOWN
 - Command = RESET
 - Transceiver type = _433_92MHZ_TRANSCEIVER
 - Firmware version = 226
 - Firmware type = Ext
 - Hardware version = 1.2
 - Undecoded packets = false
 - RFU6 packets = false
 - RFU5 packets = false
 - RFU4 packets = false
 - RFU3 packets = false
 - FineOffset / Viking (433.92) packets = false
 - Rubicson (433.92) packets = false
 - AE (433.92) packets = false
 - BlindsT1/T2/T3 (433.92) packets = false
 - BlindsT0 (433.92) packets = false
 - ProGuard (868.35 FSK) packets = false
 - FS20 (868.35) packets = false
 - La Crosse (433.92/868.30) packets = true
 - Hideki/UPM (433.92) packets = true
 - AD (433.92) packets = false
 - Mertik (433.92) packets = false
 - Visonic (315/868.95) packets = false
 - ATI (433.92) packets = false
 - Oregon Scientific (433.92) packets = true
 - Meiantech (433.92) packets = false
 - HomeEasy EU (433.92) packets = false
 - AC (433.92) packets = true
 - ARC (433.92) packets = true
 - X10 (310/433.92) packets = true
 - HomeConfort packets = false
 - KeeLoq (433.92) packets = false
2016-12-24 06:43:31.404 [ERROR] [.rfxcom.internal.RFXComBinding] - No acknowledge received from RFXCOM controller, timeout 5000ms due to java.util.concurrent.TimeoutException: waiting timed out
```

--- a/addons/binding/org.openhab.binding.rfxcom.test/src/test/java/org/openhab/binding/rfxcom/internal/messages/RFXComLighting5MessageTest.java
+++ b/addons/binding/org.openhab.binding.rfxcom.test/src/test/java/org/openhab/binding/rfxcom/internal/messages/RFXComLighting5MessageTest.java
@@ -8,10 +8,18 @@
  */
 package org.openhab.binding.rfxcom.internal.messages;
 
+import org.eclipse.smarthome.core.library.types.OnOffType;
 import org.junit.Test;
+import org.openhab.binding.rfxcom.RFXComValueSelector;
 import org.openhab.binding.rfxcom.internal.exceptions.RFXComException;
 import org.openhab.binding.rfxcom.internal.exceptions.RFXComNotImpException;
-import org.openhab.binding.rfxcom.internal.messages.RFXComBaseMessage.PacketType;
+
+import javax.xml.bind.DatatypeConverter;
+
+import static org.junit.Assert.assertEquals;
+import static org.openhab.binding.rfxcom.internal.messages.RFXComBaseMessage.PacketType.LIGHTING5;
+import static org.openhab.binding.rfxcom.internal.messages.RFXComLighting5Message.Commands.ON;
+import static org.openhab.binding.rfxcom.internal.messages.RFXComLighting5Message.SubType.IT;
 
 /**
  * Test for RFXCom-binding
@@ -21,14 +29,24 @@ import org.openhab.binding.rfxcom.internal.messages.RFXComBaseMessage.PacketType
  */
 public class RFXComLighting5MessageTest {
     @Test
-    public void checkForSupportTest() throws RFXComException, RFXComNotImpException {
-        RFXComMessageFactory.createMessage(PacketType.LIGHTING5);
+    public void basicBoundaryCheck() throws RFXComException, RFXComNotImpException {
+        RFXComTestHelper.basicBoundaryCheck(LIGHTING5);
     }
 
     @Test
-    public void basicBoundaryCheck() throws RFXComException, RFXComNotImpException {
-        RFXComTestHelper.basicBoundaryCheck(PacketType.LIGHTING5);
+    public void convertFromStateItMessage() throws RFXComException, RFXComNotImpException {
+        RFXComMessage itMessageObject = RFXComMessageFactory.createMessage(LIGHTING5);
+        itMessageObject.setDeviceId("2061.1");
+        itMessageObject.setSubType(IT);
+        itMessageObject.convertFromState(RFXComValueSelector.COMMAND, OnOffType.ON);
+        byte[] message = itMessageObject.decodeMessage();
+        String hexMessage = DatatypeConverter.printHexBinary(message);
+        assertEquals("Message is not as expected", "0A140F0000080D01010000", hexMessage);
+        RFXComLighting5Message msg = (RFXComLighting5Message) RFXComMessageFactory.createMessage(message);
+        assertEquals("SubType", IT, msg.subType);
+        assertEquals("Sensor Id", "2061.1", msg.getDeviceId());
+        assertEquals("Command", ON, msg.command);
     }
 
-    // TODO please add tests for real messages
+    // TODO please add more tests for different messages
 }

--- a/addons/binding/org.openhab.binding.rfxcom/ESH-INF/thing/lighting5.xml
+++ b/addons/binding/org.openhab.binding.rfxcom/ESH-INF/thing/lighting5.xml
@@ -31,17 +31,20 @@
 				<label>Sub Type</label>
 				<description>Specifies device sub type.</description>
 				<options>
-					<option value="LIGHTWAVERF">LightwaveRF, Siemens</option>
-					<option value="EMW100">EMW100 GAO/Everflourish</option>
-					<option value="BBSB_NEW">BBSB new types</option>
-					<option value="MDREMOTE">MDREMOTE LED dimmer</option>
-					<option value="CONRAD_RSL2">Conrad RSL2</option>
-					<option value="LIVOLO">Livolo Dimmer or On/Off 1-3</option>
-					<option value="RGB_TRC02">RGB TRC02 (2 batt)</option>
 					<option value="AOKE">Aoke Relay</option>
-					<option value="RGB_TRC02_2">RGB TRC02_2 (3 batt)</option>
+					<option value="AVANTEK">Avantek</option>
+					<option value="BBSB_NEW">BBSB new types</option>
+					<option value="CONRAD_RSL2">Conrad RSL2</option>
+					<option value="EMW100">EMW100 GAO/Everflourish</option>
 					<option value="EURODOMEST">Eurodomest</option>
-					<option value="LIVOLO_APPLIANCE">Livolo Appliance On/Off 1-10</option>
+					<option value="IT">IT</option>
+					<option value="LIGHTWAVERF">LightwaveRF, Siemens</option>
+					<option value="LIVOLO">Livolo Dimmer or On/Off 1-3</option>
+					<option value="MDREMOTE">MDREMOTE LED dimmer v106</option>
+					<option value="MDREMOTE_107">MDREMOTE v107</option>
+					<option value="MDREMOTE_108">MDREMOTE v108, EKAB-10KRF</option>
+					<option value="RGB_TRC02">RGB TRC02 (2 batt)</option>
+					<option value="RGB_TRC02_2">RGB TRC02_2 (3 batt)</option>
 				</options>
 			</parameter>
 		</config-description>

--- a/addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComLighting5Message.java
+++ b/addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComLighting5Message.java
@@ -8,8 +8,16 @@
  */
 package org.openhab.binding.rfxcom.internal.messages;
 
-import org.eclipse.smarthome.core.library.items.*;
-import org.eclipse.smarthome.core.library.types.*;
+import org.eclipse.smarthome.core.library.items.ContactItem;
+import org.eclipse.smarthome.core.library.items.DimmerItem;
+import org.eclipse.smarthome.core.library.items.NumberItem;
+import org.eclipse.smarthome.core.library.items.RollershutterItem;
+import org.eclipse.smarthome.core.library.items.SwitchItem;
+import org.eclipse.smarthome.core.library.types.DecimalType;
+import org.eclipse.smarthome.core.library.types.IncreaseDecreaseType;
+import org.eclipse.smarthome.core.library.types.OnOffType;
+import org.eclipse.smarthome.core.library.types.OpenClosedType;
+import org.eclipse.smarthome.core.library.types.PercentType;
 import org.eclipse.smarthome.core.types.State;
 import org.eclipse.smarthome.core.types.Type;
 import org.eclipse.smarthome.core.types.UnDefType;
@@ -20,11 +28,20 @@ import java.math.BigDecimal;
 import java.util.Arrays;
 import java.util.List;
 
+import static org.openhab.binding.rfxcom.internal.messages.RFXComLighting5Message.SubType.AVANTEK;
+import static org.openhab.binding.rfxcom.internal.messages.RFXComLighting5Message.SubType.BBSB_NEW;
+import static org.openhab.binding.rfxcom.internal.messages.RFXComLighting5Message.SubType.CONRAD_RSL2;
+import static org.openhab.binding.rfxcom.internal.messages.RFXComLighting5Message.SubType.EMW100;
+import static org.openhab.binding.rfxcom.internal.messages.RFXComLighting5Message.SubType.EURODOMEST;
+import static org.openhab.binding.rfxcom.internal.messages.RFXComLighting5Message.SubType.IT;
+import static org.openhab.binding.rfxcom.internal.messages.RFXComLighting5Message.SubType.LIGHTWAVERF;
+
 /**
  * RFXCOM data class for lighting5 message.
  *
  * @author Paul Hampson, Neil Renaud - Initial contribution
  * @author Pauli Anttila
+ * @author Martin van Wingerden - added support for IT and some other subtypes
  */
 public class RFXComLighting5Message extends RFXComBaseMessage {
 
@@ -37,19 +54,19 @@ public class RFXComLighting5Message extends RFXComBaseMessage {
         LIVOLO(5),
         RGB_TRC02(6),
         AOKE(7),
-        RGB_TRC02_2(7),
-        EURODOMEST(8),
-        LIVOLO_APPLIANCE(9),
+        RGB_TRC02_2(8),
+        EURODOMEST(9),
+        LIVOLO_APPLIANCE(10),
+        MDREMOTE_107(12),
+        AVANTEK(14),
+        IT(15),
+        MDREMOTE_108(16),
 
         UNKNOWN(255);
 
         private final int subType;
 
         SubType(int subType) {
-            this.subType = subType;
-        }
-
-        SubType(byte subType) {
             this.subType = subType;
         }
 
@@ -68,47 +85,62 @@ public class RFXComLighting5Message extends RFXComBaseMessage {
         }
     }
 
+    /**
+     * Note: for the lighting5 commands, some command are only supported for certain sub types and
+     * command-bytes might even have a different meaning for another sub type.
+     * <p>
+     * If no sub types are specified for a command, its supported by all sub types.
+     * An example is the command OFF which is represented by the byte 0x00 for all subtypes.
+     * <p>
+     * Otherwise the list of sub types after the command-bytes indicates the sub types
+     * which support this command with this byte.
+     * Example byte value 0x03 means GROUP_ON for IT and some others while it means MOOD1 for LIGHTWAVERF
+     */
     public enum Commands {
-        OFF(0),
-        ON(1),
-        GROUP_OFF(2),
-        MOOD1(3),
-        MOOD2(4),
-        MOOD3(5),
-        MOOD4(6),
-        MOOD5(7),
-        RESERVED1(8),
-        RESERVED2(9),
-        UNLOCK(10),
-        LOCK(11),
-        ALL_LOCK(12),
-        CLOSE_RELAY(13),
-        STOP_RELAY(14),
-        OPEN_RELAY(15),
-        SET_LEVEL(16),
-        COLOUR_PALETTE(17),
-        COLOUR_TONE(18),
-        COLOUR_CYCLE(19),
+        OFF(0x00),
+        ON(0x01),
+        GROUP_OFF(0x02, LIGHTWAVERF, BBSB_NEW, CONRAD_RSL2, EURODOMEST, AVANTEK, IT),
+        LEARN(0x02, EMW100),
+        GROUP_ON(0x03, BBSB_NEW, CONRAD_RSL2, EURODOMEST, AVANTEK, IT),
+        MOOD1(0x03, LIGHTWAVERF),
+        MOOD2(0x04, LIGHTWAVERF),
+        MOOD3(0x05, LIGHTWAVERF),
+        MOOD4(0x06, LIGHTWAVERF),
+        MOOD5(0x07, LIGHTWAVERF),
+        RESERVED1(0x08, LIGHTWAVERF),
+        RESERVED2(0x09, LIGHTWAVERF),
+        UNLOCK(0x0A, LIGHTWAVERF),
+        LOCK(0x0B, LIGHTWAVERF),
+        ALL_LOCK(0x0C, LIGHTWAVERF),
+        CLOSE_RELAY(0x0D, LIGHTWAVERF),
+        STOP_RELAY(0x0E, LIGHTWAVERF),
+        OPEN_RELAY(0x0F, LIGHTWAVERF),
+        SET_LEVEL(0x10, LIGHTWAVERF, IT),
+        COLOUR_PALETTE(0x11, LIGHTWAVERF),
+        COLOUR_TONE(0x12, LIGHTWAVERF),
+        COLOUR_CYCLE(0x13, LIGHTWAVERF),
 
         UNKNOWN(255);
 
         private final int command;
+        private final List<SubType> supportedBySubTypes;
 
         Commands(int command) {
-            this.command = command;
+            this(command, SubType.values());
         }
 
-        Commands(byte command) {
+        Commands(int command, SubType... supportedBySubTypes) {
             this.command = command;
+            this.supportedBySubTypes = Arrays.asList(supportedBySubTypes);
         }
 
         public byte toByte() {
             return (byte) command;
         }
 
-        public static Commands fromByte(int input) {
+        public static Commands fromByte(int input, SubType subType) {
             for (Commands c : Commands.values()) {
-                if (c.command == input) {
+                if (c.command == input && c.supportedBySubTypes.contains(subType)) {
                     return c;
                 }
             }
@@ -162,7 +194,7 @@ public class RFXComLighting5Message extends RFXComBaseMessage {
         sensorId = (data[4] & 0xFF) << 16 | (data[5] & 0xFF) << 8 | (data[6] & 0xFF) << 0;
         unitCode = data[7];
 
-        command = Commands.fromByte(data[8]);
+        command = Commands.fromByte(data[8], subType);
 
         dimmingLevel = data[9];
         signalLevel = (byte) ((data[10] & 0xF0) >> 4);
@@ -209,7 +241,7 @@ public class RFXComLighting5Message extends RFXComBaseMessage {
     /**
      * Convert a 0-31 scale value to a percent type.
      *
-     * @param pt
+     * @param value
      *            percent type to convert
      * @return converted value 0-31
      */

